File: 97 Interleaving String.py
class Solution:
    def isInterleave(self, s1, s2, s3):
        if len(s1) + len(s2) != len(s3):#長度不對
            return False
        if not(len(s1)) or not(len(s2)):#其中一個是空字串
            return s1 + s2 == s3

        d_s1 = [True]
        for i in range(1, len(s1)+1):#把s1和s3從頭比對連續幾個相同
            if not(d_s1[i-1]):
                break;
            d_s1.append(s1[i-1] == s3[i-1])
        while len(d_s1) < len(s1)+1:#末端補0
            d_s1.append(False)

        d_s2 = [True]
        for i in range(1, len(s2)+1):#把s2和s3從頭比對連續幾個相同
            if not(d_s2[i-1]):
                break;
            d_s2.append(s2[i-1] == s3[i-1])
        while len(d_s2) < len(s2)+1:#末端補0
            d_s2.append(False)

        dp = [d_s2]
        for i in range(1, len(d_s1)):#手動產生二維陣列
            dp.append([d_s1[i]])
            for j in range(1, len(d_s2)):
                dp[i].append((dp[i-1][j] and s3[i+j-1] == s1[i-1]) or
                             (dp[i][j-1] and s3[i+j-1] == s2[j-1]))

        return dp[len(dp)-1][len(dp[0])-1]

File: 97 Interleaving String.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        if(s1.size() + s2.size() != s3.size())
            return false;
        if(s1.empty() || s2.empty())
            return s1+s2==s3;
        bool arr[s1.size()+1][s2.size()+1];
        memset(arr, false, sizeof(arr));

        arr[0][0] = true;
        for(int i = 1;arr[i-1][0] && i < s1.size()+1;++i){
            arr[i][0] = (s1[i-1] == s3[i-1]);
        }
        for(int i = 1;arr[0][i-1] && i < s2.size()+1;++i){
            arr[0][i] = (s2[i-1] == s3[i-1]);
        }
        int i = 1, j = 1;
        for(;i < s1.size()+1;++i){
            for(j = 1;j < s2.size()+1;++j){
                arr[i][j] = (arr[i-1][j] && s3[i+j-1] == s1[i-1]) || 
                            (arr[i][j-1] && s3[i+j-1] == s2[j-1]);
            }
        }
        return arr[i-1][j-1];
    }
};